refactor: separate dev inputs

terlar opened the pull request:
This is using the flake-parts feature partitions in order to move inputs not needed by third-party consumers into a separate dev flake.

This reduces the amount of inputs that gets inherited by downstream users.

sshine commented on dev/flake-module.nix:
```diff
@@ -0,0 +1,77 @@
+    {
+      devShells.default = pkgs.mkShell {
+        buildInputs = with pkgs; [
+          terraform
```
Since we no longer `export NIXPKGS_ALLOW_UNFREE=1`, won't including `pkgs.terraform` in the dev shell lead to errors? I just tried adding it to a shell.nix earlier today, and this happened:

```
error:
       … while calling the 'derivationStrict' builtin
         at <nix/derivation-internal.nix>:37:12:
           36|
           37|   strict = derivationStrict drvAttrs;
             |            ^
           38|

       … while evaluating derivation 'nix-shell'
         whose name attribute is located at /nix/store/isfbldda5j8j6x3nbv1zim0c0dpf90v8-source/pkgs/stdenv/generic/make-derivation.nix:438:13

       … while evaluating attribute 'nativeBuildInputs' of derivation 'nix-shell'
         at /nix/store/isfbldda5j8j6x3nbv1zim0c0dpf90v8-source/pkgs/stdenv/generic/make-derivation.nix:490:13:
          489|             depsBuildBuild = elemAt (elemAt dependencies 0) 0;
          490|             nativeBuildInputs = elemAt (elemAt dependencies 0) 1;
             |             ^
          491|             depsBuildTarget = elemAt (elemAt dependencies 0) 2;

       (stack trace truncated; use '--show-trace' to show the full, detailed trace)

       error: Package ‘terraform-1.11.4’ in /nix/store/isfbldda5j8j6x3nbv1zim0c0dpf90v8-source/pkgs/applications/networking/cluster/terraform/default.nix:79 has an unfree license (‘bsl11’), refusing to evaluate.

       a) To temporarily allow unfree packages, you can use an environment variable
          for a single invocation of the nix tools.

            $ export NIXPKGS_ALLOW_UNFREE=1

          Note: When using `nix shell`, `nix build`, `nix develop`, etc with a flake,
                then pass `--impure` in order to allow use of environment variables.

       b) For `nixos-rebuild` you can set
         { nixpkgs.config.allowUnfree = true; }
       in configuration.nix to override this.

       Alternatively you can configure a predicate to allow specific packages:
         { nixpkgs.config.allowUnfreePredicate = pkg: builtins.elem (lib.getName pkg) [
             "terraform"
           ];
         }

       c) For `nix-env`, `nix-build`, `nix-shell` or any other Nix command you can add
         { allowUnfree = true; }
       to ~/.config/nixpkgs/config.nix.
```

I don't like any of those solutions, so I just switched to `pkgs.opentofu`.

This is the first time I use OpenTofu, though. Is that the way to go?

terlar replied:
I'm fine with switching to opentofu as well, that is what I personally use. But I guess it is good to have terraform support as well in the tests. Probably have tests and examples for both would be best.

sshine commented on flake.nix:
```diff
@@ -6,125 +6,74 @@
-        inputs.systems;
+      systems = import inputs.systems;
+
+      partitions.dev = {
+        extraInputsFlake = ./dev;
+        module.imports = [ ./dev/flake-module.nix ];
+      };
+
+      partitionedAttrs = {
+        apps = "dev";
+        devShells = "dev";
+        formatter = "dev";
+        templates = "dev";
+      };
```
Excellent. I have never tried to use this; I am not in a position to evaluate this yet, but I suppose the effect is obvious (smaller flake lock footprint).

terlar replied:
Yes, this is what we want to achieve. Since those inputs are used only for testing purposes. With this the downstream users of terranix won't have them added to their flake.lock.

sshine reviewed (changes requested):
Hello @terlar, and thanks for the effort.

Since I'm new to terranix and you know more about it than me, I hope we can use the review for my learning.

This PR contains a lot of line changes, and a lot of them is just reformatting whitespace. I don't mind that it does, but I'd like that all the things it does contain are stated. I'll try to iterate them.

## List of logical changes

- Removes `NIXPKGS_ALLOW_UNFREE=1` from .envrc
- Removes `--impure` from .envrc
- Makes use of `flake-parts.lib.partitions` (the main thing)
- ...I might be missing something that I overlooked due to whitespace formatting.

I will try to re-review once the formatting is addressed.

## Formatting

Reformatting a lot of whitespace makes reviewing quite time-consuming because I have to read every whitespace change and assume it's a logical change until I determine that it's not.

Reformatting can be nice, but what formatter do you use?

If we should reformat, it should probably conform to some standard, otherwise we can reformat perpetually without reaching a fixed point.

Depending on what's most convenient for you, could you either:
- Move all the whitespace formatting to a separate commit, so I can review the commit and determine that no logical changes are made, or,
- Move all the whitespace formatting to another pull request

In either case, I would like for the formatter to be known.

terlar commented:
@sshine Thank you for taking a look.

# Formatting
I split up the formatting into it's own commit. So you should be able to review the other one and not pay that much attention to the formatting. I however see that I did the formatting change last, I think it will be better to do it the other way around.

This is not using any manual formatting, it is using treefmt together with `nixfmt-rfc-style` which is kind of the official one that has been adopted by nixpkgs recently.

It was a bit confusing previously the `treefmt.toml` specified `nixpkgs-fmt` while the dev-shell added `nixfmt`. Since `nixfmt` is "deprecated" I went with `nixfmt-rfc-style`. But perhaps we can go for `nixpkgs-fmt` for now to be consistent with previous formatting.

The formatting that should be used is the packages provided by devshell and then just run `nix fmt` or `treefmt`.

# Logical changes
Since I set the following in the flake-parts:
```nix
  _module.args.pkgs = import inputs.nixpkgs {
    inherit system;
    config = {
      allowUnfree = true;
    };
  };
```

We don't have to rely on the impure hacks.

terlar commented:
@sshine I have now updated the merge request using the `nixpkgs-fmt` (also added it to the devshell). Now it should be easier to spot the changes. I have not added much, mainly lifted out the test things into the dev flake.

If you have any questions feel free to ask.

sshine commented:
Excellent. I’ll approve tomorrow morning.

terlar commented on flake.nix:
```diff
@@ -29,24 +17,36 @@
-          pkgs = import inputs.nixpkgs {
+        }:
+        {
+          _module.args.pkgs = import inputs.nixpkgs {
```
This change let's us skip the impure devShell used by direnv, since the pkgs attribute is shared among all the configuration (including the dev shell).
